mujoco_warp/_src/io.py

```diff
@@ -1303,6 +1304,547 @@ def reset_contact(
+        inputs=[tenid, m.nv, d.ten_J, ten_result_vec],
+        outputs=[m.tendon_invweight0],
+      )
+
+  @nested_kernel(module="unique", enable_backward=False)
+  def compute_cam_pos0(
+    cam_bodyid: wp.array(dtype=int),
+    cam_targetbodyid: wp.array(dtype=int),
+    cam_xpos_in: wp.array2d(dtype=wp.vec3),
+    cam_xmat_in: wp.array2d(dtype=wp.mat33),
+    xpos_in: wp.array2d(dtype=wp.vec3),
+    subtree_com_in: wp.array2d(dtype=wp.vec3),
+    cam_pos0_out: wp.array2d(dtype=wp.vec3),
+    cam_poscom0_out: wp.array2d(dtype=wp.vec3),
+    cam_mat0_out: wp.array2d(dtype=wp.mat33),
+  ):
+    worldid, camid = wp.tid()
+    cam_pos0_id = worldid % cam_pos0_out.shape[0]
+    bodyid = cam_bodyid[camid]
+    targetid = cam_targetbodyid[camid]
+    cam_xpos = cam_xpos_in[worldid, camid]
+
+    cam_pos0_out[cam_pos0_id, camid] = cam_xpos - xpos_in[worldid, bodyid]
+    if targetid >= 0:
+      cam_poscom0_out[cam_pos0_id, camid] = cam_xpos - subtree_com_in[worldid, targetid]
+    else:
+      cam_poscom0_out[cam_pos0_id, camid] = cam_xpos - subtree_com_in[worldid, bodyid]
+    cam_mat0_out[cam_pos0_id, camid] = cam_xmat_in[worldid, camid]
+
+  if m.ncam > 0:
+    wp.launch(
+      compute_cam_pos0,
+      dim=(d.nworld, m.ncam),
+      inputs=[m.cam_bodyid, m.cam_targetbodyid, d.cam_xpos, d.cam_xmat, d.xpos, d.subtree_com],
+      outputs=[m.cam_pos0, m.cam_poscom0, m.cam_mat0],
+    )
+
+  @nested_kernel(module="unique", enable_backward=False)
+  def compute_light_pos0(
+    light_bodyid: wp.array(dtype=int),
+    light_targetbodyid: wp.array(dtype=int),
+    light_xpos_in: wp.array2d(dtype=wp.vec3),
+    light_xdir_in: wp.array2d(dtype=wp.vec3),
+    xpos_in: wp.array2d(dtype=wp.vec3),
+    subtree_com_in: wp.array2d(dtype=wp.vec3),
+    light_pos0_out: wp.array2d(dtype=wp.vec3),
+    light_poscom0_out: wp.array2d(dtype=wp.vec3),
+    light_dir0_out: wp.array2d(dtype=wp.vec3),
+  ):
+    worldid, lightid = wp.tid()
+    light_pos0_id = worldid % light_pos0_out.shape[0]
+    bodyid = light_bodyid[lightid]
+    targetid = light_targetbodyid[lightid]
+    light_xpos = light_xpos_in[worldid, lightid]
+
+    light_pos0_out[light_pos0_id, lightid] = light_xpos - xpos_in[worldid, bodyid]
+    if targetid >= 0:
+      light_poscom0_out[light_pos0_id, lightid] = light_xpos - subtree_com_in[worldid, targetid]
+    else:
+      light_poscom0_out[light_pos0_id, lightid] = light_xpos - subtree_com_in[worldid, bodyid]
+    light_dir0_out[light_pos0_id, lightid] = light_xdir_in[worldid, lightid]
+
+  if m.nlight > 0:
+    wp.launch(
+      compute_light_pos0,
+      dim=(d.nworld, m.nlight),
+      inputs=[m.light_bodyid, m.light_targetbodyid, d.light_xpos, d.light_xdir, d.xpos, d.subtree_com],
+      outputs=[m.light_pos0, m.light_poscom0, m.light_dir0],
+    )
+
+  # actuator_acc0[i] = ||inv(M) * actuator_moment[i]|| - acceleration from unit actuator force
+  if m.nu > 0 and m.nv > 0:
+    act_moment_vec = wp.zeros((d.nworld, m.nv), dtype=float)
```

Comment:
if all of the array elements are initialized by kernels below we can utilize `wp.empty`